ral-facilities/operationsgateway
# DSEGOG-8 user can reorder columns

## Description
Note: this contains resizing columns functionality so probably best to review that PR first - https://github.com/ral-facilities/operationsgateway/pull/4

Adding functionality to allow the user to reorder the table columns using drag and drop functionality provided by [react-beautiful-dnd](https://github.com/atlassian/react-beautiful-dnd). The result is a nice looking, simplistic method by which the user can reorder the displayed columns without having to add or remove columns to achieve the same thing. To drag and drop, the user can click anywhere on the column header that isn't the close button or the divider. I had to be a bit clever to make sure their functionality didn't overlap.

It's not really feasible to test this functionality outside of a browser so I used this opportunity to add Cypress tests to the project. I've only done tests for this functionality so far, not made any for the previous functionality I've done yet. Currently using Cypress 9.6.0 to match DataGateway because this was the simplest to get set up with. I considered upgrading to Cypress v10 but this contains many breaking changes so I made a separate issue (stfc.atlassian.net/jira/software/projects/DSEGOG/issues/DSEGOG-33) for this to tackle later.

## Testing instructions
Experiment with adding columns to the display and then reordering them by dragging the headers. Experiment with both large and small tables to make sure the functionality works in all cases. Also make sure that the drag and drop functionality does not interfere with column resizing and closing.


## Agile board tracking
Closes [DSEGOG-8](https://stfc.atlassian.net/jira/software/projects/DSEGOG/boards/22?selectedIssue=DSEGOG-8)
